src/iceberg/type.cc

```diff
@@ -0,0 +1,312 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "iceberg/type.h"
+
+#include <format>
+#include <iterator>
+#include <stdexcept>
+
+#include "iceberg/util/formatter.h"
+
+namespace iceberg {
+
+StructType::StructType(std::vector<SchemaField> fields) : fields_(std::move(fields)) {
+  size_t index = 0;
+  for (const auto& field : fields_) {
+    auto [it, inserted] = field_id_to_index_.try_emplace(field.field_id(), index);
+    if (!inserted) {
+      throw std::runtime_error(
+          std::format("StructType: duplicate field ID {} (field indices {} and {})",
+                      field.field_id(), it->second, index));
+    }
+
+    ++index;
+  }
+}
+
+TypeId StructType::type_id() const { return TypeId::kStruct; }
+std::string StructType::ToString() const {
+  std::string repr = "struct<\n";
+  for (const auto& field : fields_) {
+    std::format_to(std::back_inserter(repr), "  {}\n", field);
+  }
+  repr += ">";
+  return repr;
+}
+std::span<const SchemaField> StructType::fields() const { return fields_; }
+std::optional<std::reference_wrapper<const SchemaField>> StructType::GetFieldById(
+    int32_t field_id) const {
+  auto it = field_id_to_index_.find(field_id);
+  if (it == field_id_to_index_.end()) return std::nullopt;
+  return fields_[it->second];
+}
+std::optional<std::reference_wrapper<const SchemaField>> StructType::GetFieldByIndex(
+    int32_t index) const {
+  if (index < 0 || index >= static_cast<int>(fields_.size())) {
+    return std::nullopt;
+  }
+  return fields_[index];
+}
+std::optional<std::reference_wrapper<const SchemaField>> StructType::GetFieldByName(
+    std::string_view name) const {
+  // TODO: what is the right behavior if there are duplicate names? (Are
```

Comment:
It seems that the spec is unclear about this so the behavior is undefined and implementation-specific.

Reply:
There seem to be a lot of spec issues...

Reply:
Please consider the Java implementation as the reference one. There we raise a `ValidationException` when we see a duplicate name:

https://github.com/apache/iceberg/blob/507e2a9e39f93356bdc0dae1521bbdc46629681c/api/src/main/java/org/apache/iceberg/types/IndexByName.java#L199-L205

> There seem to be a lot of spec issues...

@lidavidm Please elaborate

Reply:
Is this in the written spec? Or is the canonical-ness of the Java implementation documented somewhere? Just so I know what to reference

Reply:
Also see above e.g. https://github.com/apache/iceberg-cpp/pull/31#discussion_r1929559679
https://github.com/apache/iceberg-cpp/pull/31#discussion_r1928875579

possibly I overlooked something

Reply:
The spec describes the table format itself, for the behavioral things, we tend to check the Java implementation (considered the reference implementation). If we would need to formalize all this as well, then the spec would become unreadable.

Reply:
IMO, at least something like "duplicate field names are not allowed" should be part of the spec, no? I wouldn't really qualify that as "behavioral" any more than "decimal precision is limited to 38" is behavioral

Reply:
Since Iceberg comes from the database world, not having duplicate names is more or less implied to me, but feel free to raise a PR to add this to the spec itself.

I was surprised to see that this is allowed:

```python
>>> import pyarrow as pa
>>> pa.table([
...     pa.array(["foo", "bar"]),
...     pa.array(["Parrot", "Dog"])
... ], names=['animals', 'animals'])
pyarrow.Table
animals: string
animals: string
----
animals: [["foo","bar"]]
animals: [["Parrot","Dog"]]
```

Reply:
> I was surprised to see that this is allowed:

I think Arrow is a different. It is totally valid to project the same column more than once in an Arrow Table or RecordBatch. For example, returning query result via Arrow for `select animals, animals from foo.bar`;